Container resolvers: add docs, typing and tests

bernt-matthias opened the pull request:
I found it incredible hard to understand how container resolvers work (which seems important if one wants to use them in production): 

- docs in the sample config [seemed to be partially wrong and certainly incomplete](https://github.com/galaxyproject/galaxy/issues/15570)
- code is very confusing and essentially undocumented

So I took approx a week of and studied the code and compiled what I found out (hope that I got it right) in the the docs added here. There are still some TODOs where I have not found the time to investigate. But I hope its good enough for a first round of review. 

I tried to change no functionality except for the following three commits which should fix https://github.com/galaxyproject/galaxy/issues/15673

- [containers: do not mount /tmp/](https://github.com/galaxyproject/galaxy/pull/15614/commits/fffb6f80be3c962cd0f3d4568ae207c8e852c0c3)
- [refactor volume preprocessing](https://github.com/galaxyproject/galaxy/pull/15614/commits/5637129c5eb9abf07328b08765375342436a7b03)
- [remove volumes with duplicate targets](https://github.com/galaxyproject/galaxy/pull/15614/commits/b37085b50a1e2e5f7c6270000ceaab3d9dbc6977)

TODO:

- [x] more tests (suggestions very welcome). Ideally each of the statements in the docs would be backed up by a test. 
- [ ] config schema
- [x] add yaml config sample (without duplicating docs to much?)
- [ ] can the docs be integrated in the module docs? 
- [x] update xsd docs linking to https://galaxyproject.org/admin/tools/docker/
  - [x] and this hub page https://github.com/galaxyproject/galaxy-hub/blob/master/content/admin/tools/docker/index.md https://github.com/galaxyproject/galaxy-hub/issues/1961

Possible future TODOs (some things that I found odd which might go to separate PR / issues after this is done .. after tests have been added):

- `mulled_singularity` and `cached_mulled_singularity` (not sure about `build_mulled_singularity`)  currently do to use (the name of the registry or) the namespace for saving images, but store all images in `container_cache/singularity/mulled/MULLED_HASH`.
- Why is `mulled_singularity` invalidating the (file sytstem) cache manually ([here](https://github.com/galaxyproject/galaxy/blob/c517e805771cc16807dfe675075a13fe6343f01f/lib/galaxy/tool_util/deps/container_resolvers/mulled.py#L636) and not rely on the [built in mechanism](https://github.com/galaxyproject/galaxy/blob/c517e805771cc16807dfe675075a13fe6343f01f/lib/galaxy/tool_util/deps/container_resolvers/mulled.py#L138)
- registry should be configurable https://github.com/galaxyproject/galaxy/issues/10576
- `namespace="local"` seems to have the implicit assumption to mark local images (stored in dockers image cache). This seems to be used by people when building images with `planemo mull`. It would be better to force resolvers (that provide access to docker's local image registry) to use local images by using `docker-daemon://` (for singularity) and `--pull never` for docker. 
- Currently [all available images are searched for the best matching image](https://github.com/galaxyproject/galaxy/blob/c517e805771cc16807dfe675075a13fe6343f01f/lib/galaxy/tool_util/deps/container_resolvers/mulled.py#L295) .. could be inefficient for larger number of images (and frequent accesses). It might be better to just if the image for one of the constant number of possibilities is present. Or at least make [mtime cached a set](https://github.com/galaxyproject/galaxy/blob/c517e805771cc16807dfe675075a13fe6343f01f/lib/galaxy/tool_util/deps/container_resolvers/mulled.py#L145)
- Derive ExplicitSingularity* from SingularityCliResolver (in addition to ExplicitContainerResolver?)
- call _init_cache_directory in SingularityCliContainerResolver.init and not in init of derived classes?
- `container_image_cache_path` not in galaxy.yml?

Some viz for the resolvers that I found useful .. just do not want to loose it

![container_resolvers](https://user-images.githubusercontent.com/25689525/219703181-ebaee15a-8b5c-4146-bda9-b367a4ffe057.svg)

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

bernt-matthias committed:
initial commit of container resolvers documentation

bernt-matthias committed:
Typing for container resolvers

bernt-matthias committed:
add some tests for container resolvers

probably do not qualify as unit test

and maybe not a good idea

but at least useful for understanding

bernt-matthias committed:
remove container_resolvers_conf.xml

pcm32 commented on doc/source/admin/container_resolvers.rst:
```diff
@@ -0,0 +1,276 @@
+   - type: "mulled_singularity"
+     auto_install: False
+
+would define a list of three container resolvers, and the last setting an option
```
This might be a good opportunity as well to add a sample file in YAML for this config file, that reflects how the same content of the XML would be written (sometimes mapping XML attributes and children to YAML is not always straightforward, specially if one lacks an understanding of the code loading this config). This could help in the transition from XML to YAML files for people.

bernt-matthias replied:
Indeed. 

And maybe also a schema definition? But I don't understand those yet. Guess it wont be more than a mapping anyway (except maybe for the mapping resolver).

natefoo replied:
+1 to YAML. Also, the container config is generally so simple that I'm putting it inline in galaxy.yml anyway.

pcm32 commented on doc/source/admin/container_resolvers.rst:
```diff
@@ -0,0 +1,276 @@
+
+There are important differences between Galaxy's cached docker and singularity
+container resolvers. The caching mechanism essentially executes a
+``docker pull`` or ``singularity pull``, respectively. For docker this creates
```
Is the `singularity pull` currently being called? Last time I setup something relying on singularity, I had to have a mechanism of getting the tool containers apriori I think. Or maybe I did this to avoid issues with concurrent pulls when dataset collections where the first ones to request the pull of an image... not sure. Might be good to know as well if concurrent `singularity pull` calls are somehow avoided currently.

bernt-matthias replied:
Some of them do:

- `cached_explicit_singularity` triggers a pull over [here](https://github.com/galaxyproject/galaxy/blob/c517e805771cc16807dfe675075a13fe6343f01f/lib/galaxy/tool_util/deps/container_resolvers/explicit.py#L106)
- `mulled` and `mulled_singularity` trigger pulls [here](https://github.com/galaxyproject/galaxy/blob/c517e805771cc16807dfe675075a13fe6343f01f/lib/galaxy/tool_util/deps/container_resolvers/mulled.py#L592) which are then executed in the respective methods [here for `mulled`](https://github.com/galaxyproject/galaxy/blob/c517e805771cc16807dfe675075a13fe6343f01f/lib/galaxy/tool_util/deps/container_resolvers/mulled.py#L540) and [here for `mulled_singularity`](https://github.com/galaxyproject/galaxy/blob/c517e805771cc16807dfe675075a13fe6343f01f/lib/galaxy/tool_util/deps/container_resolvers/mulled.py#L632)

The pull happens when `install=True` (which is apparently the case when called during job preparation and when build is triggered from the container dependencies tab in the admin UI) ... seems that the value of `auto_intstall` is important here as well since it determines if the container description uses the URI / the path to the image (but I don't fully understand this yet).

The `cached_mulled*` only lookup the cache.

pcm32 commented on doc/source/admin/container_resolvers.rst:
```diff
@@ -0,0 +1,276 @@
+  <requirements>
+      <requirement type="package" version="0.5">foo</requirement>
+      <requirement type="package" version="1.0">bar</requirement>
+  </requirements>
```
Maybe a mention could be added here to the ability to create multi tool containers via the biocontainers UI / pull request mechanism? Or to subscribe your tool list to the CI that monitors tools that get added to the toolshed for this kind of tools?

bernt-matthias replied:
I wrote a bit here already https://github.com/galaxyproject/galaxy/blob/889f7f7413d0b3609b6829279f678e8a0e06e63f/doc/source/admin/container_resolvers.rst#building-resolver-types

It misses at the moment that also bioconda pushes to biocontainers. Maybe it should be moved to a separate subsection and also quay.org and namespaces need to be mentioned. Will put it on my todo list.

bernt-matthias committed:
cnt

nuwang commented on doc/source/admin/container_resolvers.rst:
```diff
@@ -0,0 +1,276 @@
+
+2. Mulled resolvers
+
+All mulled resolvers compute a mulled hash that desribes the reuirements and
```
```suggestion
All mulled resolvers compute a mulled hash that describes the requirements and
```

nuwang commented on doc/source/admin/container_resolvers.rst:
```diff
@@ -0,0 +1,276 @@
+  the directory where to save images.
+- ``cache_directory_cacher_type``: ``"uncached"`` (default) or ``"dir_mtime"``.
+  The singularity resolvers iterate over the contents of the cache directory. The contents
+  of the directory can be accessed uncached (thn the file listing is computed for each access)
```
```suggestion
  of the directory can be accessed uncached (in which case, the file listing is computed for each access)
```

nuwang reviewed (commented):
Nice work @bernt-matthias. This is super useful.

bernt-matthias committed:
Apply suggestions from code review

Co-authored-by: Nuwan Goonasekera <2070605+nuwang@users.noreply.github.com>

bernt-matthias committed:
doc: more intro

bernt-matthias committed:
doc cont:

add container conf sample

bernt-matthias committed:
doc

bernt-matthias committed:
doc cnt

bernt-matthias committed:
tests: add integration and fix unit tests

unit tests: by adding the now mandatory app_info argument

remove the new unit tests again (they are now implemented as integration
tests)

bernt-matthias committed:
more typing:

- made ContainerResolver.app_info required (non Optional)
  also in all other derived resolvers

  because its used non-Optional in the initialisation
  of the Container classes (where also no check for None is
  done)

- ContainerDescription.identifier required

  same reason

bernt-matthias committed:
container description: make identifier non-optional

bernt-matthias committed:
doc: cont

bernt-matthias committed:
containers: do not mount /tmp/

fixes https://github.com/galaxyproject/galaxy/issues/15673

- singularity: add `--no_mount tmp` (analogous to docker's default)
- do not bind `$TMPDIR:rw,$TMP:rw,$TEMP:rw` by default
- instead (by default )provide a writable `/tmp` via
  `$_GALAXY_JOB_TMP_DIR/:/tmp:rw` note that this dir is still also
  available as `$_GALAXY_JOB_TMP_DIR:rw`

bernt-matthias committed:
refactor volume preprocessing

to be able to cope with "default_ro" when the source directory
is given

bernt-matthias committed:
remove volumes with duplicate targets

bernt-matthias committed:
integration test runs now

due to fixes /tmp handling in containers

bernt-matthias committed:
code: post rebase cleanup

bernt-matthias committed:
test: post rebase cleanup

bernt-matthias committed:
test: restructuring

bernt-matthias committed:
doc: cnt

bernt-matthias committed:
more tests

bernt-matthias committed:
doc: continue

bernt-matthias committed:
test: cont

bernt-matthias committed:
tests: enforce container type

this is probably not excactly what is done when listing containers
in the admin UI. This is because a container type is only set via
the destination a job is executed on. That is, for listing and
building both container types are allowed, which might lead to
unwanted outcomes if the container resolver config contains resolvers
for both container types (like the default does).

gallardoalba commented on doc/source/admin/container_resolvers.rst:
```diff
@@ -0,0 +1,364 @@
+take precedence over globally defined container resolvers.
+A sample YAML file showing the default configuration which is active
+if neither a global or local configuration is given in
+`container_resolvers.yml.sample <https://github.com/galaxyproject/galaxy/tree/dev/lib/galaxy/config/sample/container_resolvers.yml.sample>`_.
```
Link not working I think.

bernt-matthias replied:
Should work when merged :)

bernt-matthias committed:
restructure tests and add tests for explicit container resolvers

bernt-matthias committed:
doc: polishing

bernt-matthias committed:
more typing

bernt-matthias committed:
follow-up restucture integration tests

bernt-matthias committed:
test: typing fixes

bernt-matthias committed:
test: slightly stricter test

bernt-matthias committed:
docs: xsd and yml sample

bernt-matthias committed:
add null container resolver

nsoranzo commented on lib/galaxy/tool_util/deps/container_resolvers/test.py:
```diff
@@ -0,0 +1,29 @@
+    from ..requirements import ContainerDescription
+
+
+class NullContainerResolver(ContainerResolver):
```
This seems currently unused?

bernt-matthias replied:
Hmm. If I grep in my current source tree I find it used in `lib/galaxy/app_unittest_utils/galaxy_mock.py`. Could you check the diff with f9116e9308?

bernt-matthias replied:
Wrong commit. It's: 894cdece7a

nsoranzo replied:
I think it's used in `test/integration/test_container_resolvers.py` when you have:

```python
    container_resolvers_config: List[Dict[str, Any]] = [
        {"type": "null"},
    ]
```

?

bernt-matthias replied:
Ah, yes. I think I used this 'trick' to create the case where the resolver can resolve...

mvdbeek replied:
In any case I don't think we should have this here just for the purpose of a test.

nsoranzo commented on doc/source/admin/container_resolvers.rst:
```diff
@@ -0,0 +1,398 @@
+Galaxy can run tools using containers using ``docker`` or ``singularity``.
+The containers can be either explicit or mulled (also called multi package containers).
+The former are given by ``<container>`` requirements pointing to a specific container.
+The later are containers built for a set of requirements of type package.
```
```suggestion
The latter are containers built for a set of requirements of type ``package``.
```

nsoranzo commented on doc/source/admin/container_resolvers.rst:
```diff
@@ -0,0 +1,398 @@
+Containers in Galaxy
+====================
+
+Galaxy can run tools using containers using ``docker`` or ``singularity``.
```
```suggestion
Galaxy can run tools inside containers using ``docker`` or ``singularity``.
```

nsoranzo commented on doc/source/admin/container_resolvers.rst:
```diff
@@ -0,0 +1,398 @@
+of single packages simply the package name and version are used instead of the hashes,
+e.g. ``ucsc-liftover:357--h446ed27_4``.
+
+The bioconda and the Galaxy project provide infrastructure to create mulled
```
```suggestion
Bioconda and the Galaxy project provide infrastructure to create mulled
```

nsoranzo commented on doc/source/admin/container_resolvers.rst:
```diff
@@ -0,0 +1,398 @@
+   repository. Mulled containers are added automatically to this repository for all tools
+   in tool repositories that are crawled by the 
+   `planemo monitor <https://github.com/galaxyproject/planemo-monitor>`_ repository
+   (which includes for instance IUC and many other tool repositories).
```
```suggestion
   (which includes for instance tools-iuc and several other tool repositories).
```

nsoranzo commented on doc/source/admin/container_resolvers.rst:
```diff
@@ -0,0 +1,398 @@
+`require_container <https://github.com/galaxyproject/galaxy/blob/0742d6e27702c60d1b8fe358ae03a267e3f252c3/lib/galaxy/config/sample/job_conf.sample.yml#L528>`_
+the job fails in this case.
+
+Besides determining a container description some container resolvers
```
```suggestion
Besides determining a container description, some container resolvers
```

nsoranzo commented on doc/source/admin/container_resolvers.rst:
```diff
@@ -0,0 +1,398 @@
+the job fails in this case.
+
+Besides determining a container description some container resolvers
+also cache/build containers.
```
```suggestion
also cache and/or build containers.
```

nsoranzo commented on doc/source/admin/container_resolvers.rst:
```diff
@@ -0,0 +1,398 @@
+
+Note that, for the execution with ``singularity`` Galaxy relies mostly on
+docker containers that are either executed directly or are converted
+to singularity images. An exception are for instance explicit container
```
```suggestion
to singularity images. An exception is for instance explicit container
```

nsoranzo commented on doc/source/admin/container_resolvers.rst:
```diff
@@ -0,0 +1,398 @@
+"""""""""""""""""""""""
+
+While non-cached resolvers will yield a container description pointing to an online
+available docker container cached resolvers will store container images on disk and
```
```suggestion
available docker container, cached resolvers will store container images on disk and
```

nsoranzo commented on doc/source/admin/container_resolvers.rst:
```diff
@@ -0,0 +1,398 @@
+   Using a cached docker resolver has no additional value on distributed compute
+   systems since the cache is only available locally. 
+   Therefore an additional ``docker inspect ... ; [ $? -ne 0 ] && docker pull ...``
+   command is used in each job script. Thereby a the container will be cached
```
```suggestion
   command is used in each job script. Thereby a container will be cached
```

nsoranzo commented on doc/source/admin/container_resolvers.rst:
```diff
@@ -0,0 +1,398 @@
+   command is used in each job script. Thereby a the container will be cached
+   after the tool run even if no cached container resolver was used.
+   Clearly admins need to take care of docker caches of the main and compute nodes.
+   For distributed compute systems built in techniques of docker may be useful:
```
```suggestion
   For distributed compute systems, built-in techniques of docker may be useful:
```

nsoranzo commented on doc/source/admin/container_resolvers.rst:
```diff
@@ -0,0 +1,398 @@
+
+The resolve function is called when 
+
+1. listing the the container tab in the dependency admin UI (using ``api/container_resolvers/toolbox``)
```
```suggestion
1. listing the container tab in the dependency admin UI (using ``api/container_resolvers/toolbox``)
```

nsoranzo commented on doc/source/admin/container_resolvers.rst:
```diff
@@ -0,0 +1,398 @@
+
+.. note::
+
+    In contrast to the uncached explicit resolver the uncached mulled resolvers
```
```suggestion
    In contrast to the uncached explicit resolver, the uncached mulled resolvers
```

nsoranzo commented on doc/source/admin/container_resolvers.rst:
```diff
@@ -0,0 +1,398 @@
+  Applies only to the resolvers listed in `Additional resolver types`_.
+- ``auto_install``: defaults to ``True``. 
+  Applies to ``mulled``, ``mulled_singularity``, ``build_mulled``, and ``build_mulled_singularity``.
+  For the non-building resolvers this controls if a contained description pointing to the
```
```suggestion
  For the non-building resolvers this controls if a container description pointing to the
```

nsoranzo commented on doc/source/admin/container_resolvers.rst:
```diff
@@ -0,0 +1,398 @@
+
+- ``cache_directory``: applies to singularity container resolvers that allow to
+  cache images and sets the directory where to save images.
+  If not set containers are saved in ``"database/container_cache/singularity/[explicit|mulled]"``.
```
```suggestion
  If not set, containers are saved in ``"database/container_cache/singularity/[explicit|mulled]"``.
```

nsoranzo commented on doc/source/admin/container_resolvers.rst:
```diff
@@ -0,0 +1,398 @@
+  or cached (then the listing is computed only if the mtime of the cache dir changes and on first access).
+  (applies to all singularity resolvers that can cache images, except explicit_singularity)
+
+Note on the built in caching capabilities of singularity and docker
```
```suggestion
Note on the built-in caching capabilities of singularity and docker
```

nsoranzo commented on doc/source/admin/container_resolvers.rst:
```diff
@@ -0,0 +1,398 @@
+Note on the built in caching capabilities of singularity and docker
+-------------------------------------------------------------------
+
+It is important to note that docker as well as singularity have their own builtin
```
```suggestion
It is important to note that docker as well as singularity have their own built-in
```

nsoranzo commented on doc/source/admin/container_resolvers.rst:
```diff
@@ -0,0 +1,398 @@
+It is important to note that docker as well as singularity have their own builtin
+caching mechanism.
+
+In case of docker a ``docker pull`` (e.g. executed from a container resolver) or
```
```suggestion
In case of docker, a ``docker pull`` (e.g. executed from a container resolver) or
```

nsoranzo commented on doc/source/admin/container_resolvers.rst:
```diff
@@ -0,0 +1,398 @@
+In case of docker a ``docker pull`` (e.g. executed from a container resolver) or
+``docker run`` (e.g. executed on the compute node running the job) will add the
+image to the **local** image cache.
+Galaxy's docker container resolvers rely on docker's built it image cache,
```
```suggestion
Galaxy's docker container resolvers rely on docker's built-in image cache,
```

nsoranzo commented on doc/source/admin/container_resolvers.rst:
```diff
@@ -0,0 +1,398 @@
+   docker container resolvers creates these image tarballs.
+
+Also singularity has its own caching mechanism and caches by default to ``$HOME/.singularity``.
+It may be cleaned regularly using ``singularity cache`` or be disabled by using the
```
```suggestion
It can be cleaned regularly using the ``singularity cache`` command, or disabled by using the
```

nsoranzo commented on doc/source/admin/container_resolvers.rst:
```diff
@@ -0,0 +1,398 @@
+
+Also singularity has its own caching mechanism and caches by default to ``$HOME/.singularity``.
+It may be cleaned regularly using ``singularity cache`` or be disabled by using the
+``SINGULARITY_DISABLE_CACHE``. Environment variable.
```
```suggestion
``SINGULARITY_DISABLE_CACHE`` environment variable.
```

nsoranzo commented on doc/source/admin/jobs.md:
```diff
@@ -111,10 +111,13 @@ In the case of Docker, containers are run using **sudo** unless ``<param id="doc
+using a ``<container>`` tag or a mulled (multi package) container is implied for the set 
+of packages specified by the tool's ``<requirement>`` tags. 
+In either case the container to be used is determined using container resolvers that can be specified
+globally for an instance of locally per instrance, see [Container resolvers](container_resolvers.md).
```
This line is unclear.

bernt-matthias commented on doc/source/admin/jobs.md:
```diff
@@ -111,10 +111,13 @@ In the case of Docker, containers are run using **sudo** unless ``<param id="doc
+using a ``<container>`` tag or a mulled (multi package) container is implied for the set 
+of packages specified by the tool's ``<requirement>`` tags. 
+In either case the container to be used is determined using container resolvers that can be specified
+globally for an instance of locally per instrance, see [Container resolvers](container_resolvers.md).
```
I think I wanted to write:

```suggestion
globally for an instance and/or per execution environment, see [Container resolvers](container_resolvers.md).
```

Does this make more sense?

I also realized that I often wrote compute environment instead of execution environment .. I will fix this.

I just have to note once that I don't really like that destinations have been renamed to execution environments. IMO the not (really clearly) speaking name was not the real problem, but the lack of definition what a destination/execution environment is. This is still missing (but I have to admit that the name is a bit better) and we have inconsistent naming in docs and code :( .. maybe it's my formal/math background that suggests me that a name is not important as long as it is defined and used consistently.

nsoranzo replied:
> Does this make more sense?

Yes, thanks, can you please make the change?

bernt-matthias committed:
Apply suggestions from code review

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

bernt-matthias committed:
fix compute -> execution environment

nsoranzo commented on test/integration/test_container_resolvers.py:
```diff
@@ -0,0 +1,1211 @@
+                {"id": "upload1", "environment": "local_upload"},
+            ],
+        }
+        # config["job_config"]["execution"]["environments"]["local_docker"]["require_container"] = True
```
```suggestion
```

nsoranzo commented on test/integration/dockerized_job_conf.yml:
```diff
@@ -10,12 +10,14 @@ execution:
     local_docker:
       runner: local
       docker_enabled: true
+      # require_container: true
```
Are the comments in this and the following file needed?

bernt-matthias replied:
I remember that I played around with them. Lets just try to enable them and see what fails. I think this makes the tests stronger since it ensures that the tests really use containers and do not fall back to conda (which is the default).

```suggestion
      require_container: true
```

nsoranzo commented on doc/source/admin/jobs.md:
```diff
@@ -111,10 +111,13 @@ In the case of Docker, containers are run using **sudo** unless ``<param id="doc
+using a ``<container>`` tag or a mulled (multi package) container is implied for the set 
+of packages specified by the tool's ``<requirement>`` tags. 
+In either case the container to be used is determined using container resolvers that can be specified
+globally for an instance of locally per instance, see [Container resolvers](container_resolvers.md).
```
```suggestion
globally for an instance of locally per instance, see [Container resolvers](container_resolvers).
```

nsoranzo commented on lib/galaxy/config/sample/galaxy.yml.sample:
```diff
@@ -710,7 +710,7 @@ galaxy:
   # If this is set to None, the default container resolvers loaded is
   # determined by enable_mulled_containers. For available options see
-  # config/container_resolvers_conf.xml.sample.
+  # https://docs.galaxyproject.org/en/master/admin/container_resolvers.rst.
```
```suggestion
  # https://docs.galaxyproject.org/en/master/admin/container_resolvers.html
```

nsoranzo commented on lib/galaxy/config/schemas/config_schema.yml:
```diff
@@ -570,7 +570,7 @@ mapping:
           this is set to None, the default container resolvers loaded is
           determined by enable_mulled_containers.
-          For available options see config/container_resolvers_conf.xml.sample.
+          For available options see https://docs.galaxyproject.org/en/master/admin/container_resolvers.rst.
```
```suggestion
          For available options see https://docs.galaxyproject.org/en/master/admin/container_resolvers.html
```

nsoranzo commented on lib/galaxy/tool_util/deps/views.py:
```diff
@@ -331,36 +350,31 @@ def clean(self, index=None, **kwds):
         tool = self._app.toolbox.tools_by_id[tool_id]
 
+        for container in tool.containers:
+            print(f"ContainerResolutionView.resolve {tool} {container.to_dict()}")
```
Debugging statement?

```suggestion
```

bernt-matthias commented on test/integration/dockerized_job_conf.yml:
```diff
@@ -10,12 +10,14 @@ execution:
       container_resolvers:
         - type: fallback
           identifier: 'quay.io/biocontainers/bwa:0.7.15--0'
+      # require_container: true
```
```suggestion
      require_container: true
```

bernt-matthias commented on test/integration/singularity_job_conf.yml:
```diff
@@ -12,6 +12,7 @@ execution:
       singularity_enabled: true
       # Since tests run in /tmp/ , we apparently need to forbid the default mounting of /tmp
       singularity_run_extra_arguments: '--no-mount tmp'
+      # require_container: true
```
```suggestion
      require_container: true
```

bernt-matthias committed:
require_container in tests

i.e. do not allow fallback to conda

bernt-matthias committed:
Apply suggestions from code review

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

nsoranzo commented on doc/source/admin/jobs.md:
```diff
@@ -111,10 +111,13 @@ In the case of Docker, containers are run using **sudo** unless ``<param id="doc
+using a ``<container>`` tag or a mulled (multi package) container is implied for the set 
+of packages specified by the tool's ``<requirement>`` tags. 
+In either case the container to be used is determined using container resolvers that can be specified
+globally for an instance of locally per instance, see [Container resolvers](container_resolvers).
```
```suggestion
globally for an instance and/or per execution environment, see [Container resolvers](container_resolvers).
```

bernt-matthias committed:
Update doc/source/admin/jobs.md

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

bernt-matthias committed:
Apply suggestions from code review

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

bernt-matthias committed:
replace container resolver config xml by yml

nsoranzo commented on lib/galaxy/config/sample/container_resolvers.yml.sample:
```diff
@@ -0,0 +1,139 @@
+# Sample file for container resolvers. Resolvers that are may be used by default
```
```suggestion
# Sample file for container resolvers. Resolvers that are potentially used by default
```

nsoranzo commented on lib/galaxy/config/sample/container_resolvers.yml.sample:
```diff
@@ -0,0 +1,139 @@
+  namespace: biocontainers
+  # In addition to the arguments of `mulled` there are cache_directory
+  # and cache_directory_cacher_type. See the description at `cached_explicit_mulled`
+  # and note the minor different in the default for the directory
```
```suggestion
  # and note the minor difference in the default for the directory
```

nsoranzo commented on lib/galaxy/config/sample/container_resolvers.yml.sample:
```diff
@@ -0,0 +1,139 @@
+- type: mulled_singularity
+  namespace: biocontainers
+  # In addition to the arguments of `mulled` there are cache_directory
+  # and cache_directory_cacher_type. See the description at `cached_explicit_mulled`
```
`cached_explicit_mulled` doesn't seem to exist?

bernt-matthias replied:
```suggestion
  # and cache_directory_cacher_type. See the description at `cached_explicit_singularity`
```

nsoranzo commented on lib/galaxy/config/sample/container_resolvers.yml.sample:
```diff
@@ -0,0 +1,139 @@
+#-type: requires_galaxy_environment_singularity
+  #identifier: A_VALID_CONTAINER_IDENTIFIER
+
+# mapping container resolver allows to specify a list of mappings from map tools
```
```suggestion
# The mapping container resolver allows to specify a list of mappings from tools
```

nsoranzo commented on lib/galaxy/config/sample/container_resolvers.yml.sample:
```diff
@@ -0,0 +1,139 @@
+  #identifier: A_VALID_CONTAINER_IDENTIFIER
+
+# mapping container resolver allows to specify a list of mappings from map tools
+# (tool_id) to containers (identifier).
```
```suggestion
# (tool_id) to containers (type and identifier).
```

nsoranzo commented on lib/galaxy/tool_util/deps/container_classes.py:
```diff
@@ -71,59 +90,185 @@
+        self.container_type = container_type
+
+    @classmethod
+    def parse_volume_str(cls, rawstr: str) -> Tuple[str, str, str]:
```
```suggestion
    @staticmethod
    def parse_volume_str(rawstr: str) -> Tuple[str, str, str]:
```

nsoranzo reviewed (approved):
Let's get this in before tomorrow's freeze.

bernt-matthias committed:
Apply suggestions from code review

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

mvdbeek commented on lib/galaxy/tool_util/deps/container_classes.py:
```diff
@@ -138,49 +283,38 @@ def preprocess_volumes(volumes_raw_str, container_type):
+        volume.mode = mode
+
+    # remove duplicate directories
+    targets: Dict[str, Volume] = dict()
```
is this whole section here not equivalent to
```python
return list(set(str(v) for v in volumes))
```
?

bernt-matthias replied:
I wanted to keep the code as (much as possible as) it is before we have 

Not sure. It's `volumes[i].target` that is used for deciding if a volume is deleted.

mvdbeek replied:
But you can't get a different string representation of a Volume instance for the same target, right ?

mvdbeek commented on lib/galaxy/tool_util/deps/container_classes.py:
```diff
@@ -221,21 +356,21 @@ def add_var(name, value):
                 defaults += ",$tmp_directory:/tmp:rw"
             else:
-                defaults += ",$_GALAXY_JOB_TMP_DIR:rw,$TMPDIR:rw,$TMP:rw,$TEMP:rw"
+                defaults = "$_GALAXY_JOB_TMP_DIR:rw,$_GALAXY_JOB_TMP_DIR/:/tmp:rw"
```
Where's the `/tmp` coming from here ?

bernt-matthias replied:
This should be related to:

> I tried to change no functionality except for the following three commits which should fix https://github.com/galaxyproject/galaxy/issues/15673

mvdbeek commented on lib/galaxy/config/sample/container_resolvers.yml.sample:
```diff
@@ -0,0 +1,139 @@
+  namespace: biocontainers
+- type: cached_mulled
+  namespace: local
+  # This container resolver has two arguments
+  # the resolver will check the image cache for `quay.io/NAMESPACE/containername`
+  # so if one (or another container resolver) did a
+  # `docker pull quay.io/NAMESPACE/containername` then the image should be found
+  # Note the local namespace is currently "abused" to refer to locally built
+  # images (since there is no quay.io/local/)
+  #namespace: biocontainers
+  # the version of the mulled hashing function (v2/v1)
+  #hash_func: v2
```
I don't think that's understandable or even really configurable in a useful way ... maybe remove this ? No-one should need to mess with `hash_func`, and I doubt any namespace that is not biocontainers will work (but if you want to test this out that'd be welcome of course)

mvdbeek commented on lib/galaxy/config/sample/container_resolvers.yml.sample:
```diff
@@ -0,0 +1,139 @@
+  namespace: biocontainers
+- type: cached_mulled_singularity
+  namespace: local
+  # NOTE that the default container config specifies the namespace, but it is
```
I don't know what you mean by redundant

bernt-matthias replied:
I mean that `cached_mulled_singularity` does the same independent of the `namespace`. It just does not use it. Therefore they are redundant. 

See, https://github.com/galaxyproject/galaxy/pull/15623 .. but we did not agree on this before.

mvdbeek commented on lib/galaxy/config/sample/container_resolvers.yml.sample:
```diff
@@ -0,0 +1,139 @@
+  # the method for caching directory listings (not the method for image caching)
+  # can be uncached, dir_mtime (the latter only determines the directory listing
+  # if the mtime of the directory changed)
+  # cache_directory_cacher_type: uncached
```
Can you make that the default ? No need to throw hurdles in the way of admins.

bernt-matthias replied:
I think it is the default.

mvdbeek replied:
I mean the cached version.

mvdbeek commented on lib/galaxy/config/sample/container_resolvers.yml.sample:
```diff
@@ -0,0 +1,139 @@
+  # cache_directory_cacher_type: uncached
+  #
+  # the version of the mulled hashing function (v2/v1)
+  #hash_func: v2
```
Again seems unnecessary

mvdbeek commented on lib/galaxy/config/sample/container_resolvers.yml.sample:
```diff
@@ -0,0 +1,139 @@
+  # the version of the mulled hashing function (v2/v1)
+  #hash_func: v2
+
+# Build a mulled container description to quay.io/NAMESPACE/MULLED_HASH where the
```
This seems to come a bit out of nowhere, was this meant to go into
the "Building container resolvers" section 

Or maybe I don't understand what you mean by `Build a mulled container description to quay.io/NAMESPACE/MULLED_HASH` ?

bernt-matthias replied:
I just wanted to highlight  that this will be the name of the container (even if it is not downloaded from quay.io).

mvdbeek commented on lib/galaxy/config/sample/container_resolvers.yml.sample:
```diff
@@ -0,0 +1,139 @@
+  # which is also quay.io/NAMESPACE/MULLED_HASH in case of docker
+  #auto_install: true
+
+# Nearly the same as mulled, but
```
I think a crucial point missing here is that if there's no prefix we're talking about docker.

mvdbeek commented on lib/galaxy/config/sample/container_resolvers.yml.sample:
```diff
@@ -0,0 +1,139 @@
+- type: mulled_singularity
+  namespace: biocontainers
+  # In addition to the arguments of `mulled` there are cache_directory
+  # and cache_directory_cacher_type. See the description at `cached_explicit_singularity`
```
That is so much of an implementation detail ... see above

bernt-matthias replied:
Agreed.

mvdbeek commented on lib/galaxy/config/sample/container_resolvers.yml.sample:
```diff
@@ -0,0 +1,139 @@
+# ----------------------------
+#
+# The following uncommented container resolvers are included in the default
+# if ``docker`` is available
```
Below it seems `build_mulled` and `build_mulled_singularity` are enabled by default ? That can't be right ?!

bernt-matthias replied:
They are, if docker is available: https://github.com/galaxyproject/galaxy/blob/e12c7993f1ef9c2a0777e979ea16de3085d99203/lib/galaxy/tool_util/deps/containers.py#L338

mvdbeek replied:
Hmm, another weird default. I guess there's no easy way to know if we need singularity images to be built.

mvdbeek commented on lib/galaxy/tool_util/deps/container_resolvers/__init__.py:
```diff
@@ -44,7 +43,7 @@ class ContainerResolver(Dictifiable, metaclass=ABCMeta):
     read_only = True  # not used for containers, but set for when they are used like dependency resolvers
 
-    def __init__(self, app_info: Optional["AppInfo"] = None, **kwds) -> None:
+    def __init__(self, app_info: "AppInfo", **kwds) -> None:
```
I don't think we should break the type annotation for the published package.

bernt-matthias replied:
We can try to revert it, but I guess this might imply changes at other places.

mvdbeek commented on lib/galaxy/tool_util/deps/container_resolvers/explicit.py:
```diff
@@ -64,22 +72,24 @@ def resolve(self, enabled_container_types, tool_info, **kwds):
-
-    def _init_cache_directory(self):
+        cache_directory_path = kwargs.get("cache_directory")
+        if not cache_directory_path:
```
`if not cache_directory_path and self.app_info:`

bernt-matthias replied:
Guess this is one of the changes that would be implied. Maybe it's not more than this.

mvdbeek commented on lib/galaxy/tool_util/deps/requirements.py:
```diff
@@ -182,15 +182,15 @@ class ToolRequirementsException(Exception):
     def __init__(
         self,
-        identifier: Optional[str] = None,
+        identifier: str,
```
Don't think we should break this.

mvdbeek commented on doc/source/admin/container_resolvers.rst:
```diff
@@ -0,0 +1,398 @@
+``docker pull`` or ``singularity pull``, respectively. For docker this creates
+an entry in the docker image cache (on the local node) whereas for
+singularity an image file is created in the specified ``cache_directory``.
+On distributed systems ``cache_directory`` needs to be accessible on all
```
but it doesn't need to be on shared storage, right ?

mvdbeek commented on doc/source/admin/container_resolvers.rst:
```diff
@@ -0,0 +1,398 @@
+
+.. note::
+
+   Using a cached docker resolver has no additional value on distributed compute
```
For starters the resolution is a lot faster, you should definitely not disable it, which one might think if they were to read this.

mvdbeek commented on test/unit/tool_util/test_container_description.py:
```diff
@@ -10,8 +10,7 @@
         ("/Cache/mUlled-abc", "/Cache/mulled-abc"),
-        ("", None),
-        (None, None),
+        ("", ""),
```
That seems like a regression

mvdbeek commented on test/integration/test_containerized_jobs.py:
```diff
@@ -10,38 +10,56 @@
+# define an environment (local_singularity) for local execution with singularity enabled
 SINGULARITY_JOB_CONFIG_FILE = os.path.join(SCRIPT_DIRECTORY, "singularity_job_conf.yml")
+
 EXTENDED_TIMEOUT = 120
```
This is much too long for the number of tests we're doing here.

mvdbeek replied:
A rough comparison with the the integration test run for the merge commit (https://github.com/galaxyproject/galaxy/actions/runs/5327223927/) and a recent run on dev before these changes (https://github.com/galaxyproject/galaxy/actions/runs/5294112030/) seems to suggest that's 20 to 30 minutes per job.

bernt-matthias replied:
Should be mostly due to the tests for the `*build*` resolvers.
